Allow core to whitelist apps for a user

sharidas opened the pull request:
This changeset allows core to whitelist apps for
a user. This way core gets controll over the apps
which a user can access or not. But this data needs
to be provided either by the app or the caller of
specific method in the IRegulateAppAcces to set.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
In this PR, I have tried to grab the information about access of apps for a user. The idea here is to have the data available for core. So lets say for example the SettingsManager resides in core. And SettingsManager cannot know what are the apps whitelisted for a user, unless the data is shared with core. So any app(s) which requires to share this data to core, should use this API. So that core has this information.

The data is saved in the array where user id is the key and the apps are the list. So we get the details per user.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
The whitelisted apps details should be available to core. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Enable guests and customgroups app
- Create a guest user
- Now login as guest user and navigate to personal settings page
- customgroups ( by default ) will not appear in the list.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

jvillafanez commented on lib/private/App/AppManager.php:
```diff
@@ -160,9 +160,13 @@ public function getInstalledApps() {
-		});
+		$regulatedAppForUser = false;
+		if ($user !== null) {
+			$regulatedAppForUser = \OC::$server->getRegulateAppAccessService()->getWhitelistedAppsForUser($user);
```
This should be injected. In addition, taking into account that the feature is deeply attached to the appManager, it might be interesting not to make the service "publicly available" but access to it through the appManager. I mean, something like `$appManager->getAccessService()`.

jvillafanez commented on lib/private/Server.php:
```diff
@@ -907,6 +908,10 @@ public function __construct($webRoot, \OC\Config $config) {
 			return $this;
 		});
+
+		$this->registerService('RegulateAppAccessService', function (Server $c) {
```
As said above, I'd rather not to expose this service directly but through the appManager.

jvillafanez commented on lib/private/App/RegulateAppAccess.php:
```diff
@@ -0,0 +1,89 @@
+	 * @return bool
+	 * @since 10.3.1
+	 */
+	public function setWhitelistedAppsForUser(IUser $user, $apps) {
```
Taking into account that the main goal is to use this for the guests apps, how is the guests app expected to use this?
Note that the list of apps can change anytime.

sharidas committed:
Allow core to whitelist apps for a user

This changeset allows core to whitelist apps for
a user. This way core gets controll over the apps
which a user can access or not. But this data needs
to be provided either by the app or the caller of
specific method in the IRegulateAppAcces to set.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

jvillafanez commented on lib/private/App/AppManager.php:
```diff
@@ -82,6 +82,8 @@ class AppManager implements IAppManager {
 	/** @var IConfig */
 	private $config;
+	/** @var RegulateAppAccess */
+	private $regulateAppAccess;
```
this needs a more meaningful name.

jvillafanez commented:
Note that we need persistence in the service.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,175 @@
+		 * Check if the user is available in different groups. If so grab the apps
+		 * user has access to these groups.
+		 */
+		$groupIds = $this->groupManager->getUserGroups($user);
```
double-check the performance with LDAP and a user in a lot of groups. This needs to be blazing fast because it will affect to all requests.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,175 @@
+		if (\is_array($groupIds) && !empty($groupIds)) {
+			foreach ($groupIds as $groupId) {
+				if (isset($this->appsAvailableForGroups[$groupId->getGID()])) {
+					$groupWhitelistedApps = \array_merge($groupWhitelistedApps, $this->appsAvailableForGroups[$groupId->getGID()]);
```
we'll have to check the performance on this one too.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,175 @@
+	 * @return bool|array get the array if whitelisted apps exist for it else false.
+	 * @since 10.3.1
+	 */
+	public function getWhitelistedAppsForGroup($gid) {
```
I'd use an `IGroup` for consistency with the other call.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,196 @@
+	 * @param array $apps
+	 * @return bool true if whitelisted apps are set else false is returned
+	 */
+	public function setWhitelistedAppsForGroup(string $group, $apps = []) {
```
Better use `IGroup` as type hint. This way we can ignore the `!empty($group)` check.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,196 @@
+	 * @return bool
+	 * @since 10.3.1
+	 */
+	public function setWhitelistedAppsForUser($uid, $apps = []) {
```
Better use `IUser` as type hint to remove the check below

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,196 @@
+	 * @since 10.3.1
+	 */
+	public function getWhitelistedAppsForUser(IUser $user) {
+		if ($user === null) {
```
I don't think this is possible to happen.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,196 @@
+	 * @since 10.3.1
+	 */
+	public function getWhitelistedAppForUserExcludingGroups(IUser $user) {
+		if ($user !== null) {
```
`$user === null` shouldn't happen

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,196 @@
+	 * @since 10.3.1
+	 */
+	public function getWhitelistedAppsForGroup(IGroup $group) {
+		if ($group === null) {
```
This shouldn't happen

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,267 @@
+	 */
+	public function updateWhitelistedAppsForGroup(IGroup $group, $apps = []) {
+		$qb = $this->connection->getQueryBuilder();
+		$qb->update('appaccess')
```
Check for "upsert". I'm not sure about the method signature, but the expectation is that it updates the row if exists and insert a new one if it doesn't exists. We can also remove this `updateWhitelistedAppsForGroup` method then.

jvillafanez commented on db_structure.xml:
```diff
@@ -1646,4 +1646,56 @@
 
 	</table>
 
+	<table>
```
I think it's better to use migrations.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,267 @@
+	 * @since 10.3.1
+	 */
+	public function setWhitelistedAppsForUser(IUser $user, $apps = []) {
+		if ($user !== null) {
```
user can't be null

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,267 @@
+	 *
+	 * @param ILogger $logger
+	 */
+	public function __construct(ILogger $logger, IDBConnection $connection = null, IGroupManager $groupManager = null) {
```
both connection and groupManager can be null, but there is no check for null. Either it's impossible that the connection is null and the "null" can be removed, or this will crash in some scenario.
Note that nullable dependencies are kind of bad, so if it's needed include a comment explaining the need.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,267 @@
+			$qb->select('apps')
+				->from('appaccess')
+				->where($qb->expr()->in('accessor', $qb->createPositionalParameter($groupGids, IQueryBuilder::PARAM_STR_ARRAY)));
+			foreach ($qb->execute()->fetchAll() as $value) {
```
`fetchAll` is usually a bad practice because you don't know the amount of information that can be loaded into memory. Either ensure the amount of information is reasonably small or document what is the expected maximum limit.

jvillafanez commented on db_structure.xml:
```diff
@@ -1646,4 +1646,56 @@
+
+			<index>
+				<name>appaccess_accessor</name>
+				<unique>true</unique>
```
this isn't right. Having an "admin" user and another "admin" group is common

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,267 @@
+			->from('appaccess')
+			->where($qb->expr()->eq('accessor', $qb->expr()->literal($group->getGID())))
+			->andWhere($qb->expr()->eq('accessor-identifier', $qb->expr()->literal('1')));
+		$result = $qb->execute()->fetch();
```
always close DB cursor when you're done in order to release memory.

jvillafanez commented:
We might need some in-memory caching. I've checked with 10k LDAP users and 1k LDAP groups, with only one whitelisted app for one group. The user who belongs to all the groups has a delay of 0.5 secs on average per request (sometimes even some seconds).
Note that this is over a local network connection. Performance might be worse if the LDAP server is outside. On the other hand, it shouldn't be too bad under normal circumstances (a user isn't expected to be member of more than 20 groups).

There is also some weird things to investigate: the third request to the `getEnabledAppsForUser` method is the one that hits the 0.5 secs, the first and second are faster, but it seems they bypass this new access restrictions somehow (3rd call in my case is https://github.com/owncloud/core/blob/master/ocs/v1.php#L58)

sharidas commented:
Before I was updating the latest changeset, there was a thought in my mind, should I remove the caching ( the one we had in our previous versions ). Because if we keep caching, then we can reduce the hits to db.

Do you also recommend to cache `$this->groupManager->getUserGroups($user);` ( i.e, caching of users in groups)

jvillafanez commented on core/Migrations/Version20191030124642.php:
```diff
@@ -0,0 +1,88 @@
+				]
+			);
+
+ 			$appAccessTable->setPrimaryKey(['id']);
```
It's unlikely we'll need indexes in this table (it's only planned for the guests group, and I don't think we'll need much more information), but I think we should include them as a general good practice

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,283 @@
+	private $userAccessAppList = [];
+
+	/** @var array */
+	private $groupAccessAppList = [];
```
Remove the caching for now. I think it's better to ensure the code works properly, so the easier to follow the better.
We can add caching later on top of this class if it's really needed.

sharidas replied:
Done.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,283 @@
+			->where($qb->expr()->eq('accessor', $qb->expr()->literal($user->getUID())))
+			->andWhere($qb->expr()->eq('accessoridentifier', $qb->expr()->literal(0)));
+		$statement = $qb->execute();
+		$result = $$statement->fetch();
```
`$$statement` doesn't look good

sharidas replied:
Removed.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,283 @@
+			$qb = $this->connection->getQueryBuilder();
+			$qb->select('apps')
+				->from('appaccess')
+				->where($qb->expr()->in('accessor', $qb->createPositionalParameter($userGroupIds, IQueryBuilder::PARAM_STR_ARRAY)));
```
this is buggy.
* "admin" user, member of "admin" group -> all apps whitelisted
* "admin" group -> "app1" and "app2" whitelisted
* "adminSubstitute" user, member of "admin" group -> "app3" and "app4" whitelisted

The "adminSustitute" should have only access to apps 1-4, but he will have access to all the apps with that query because it will get the whitelist from the "admin" user too instead of just the "admin" group.

sharidas replied:
Done

jvillafanez commented on lib/private/App/DummyAppAccessService.php:
```diff
@@ -0,0 +1,161 @@
+	 * @param IGroupManager|null $groupManager
+	 * @since 10.3.1
+	 */
+	public function __construct(ILogger $logger, IDBConnection $connection = null, IGroupManager $groupManager = null) {
```
This isn't needed. Plain constructor is enough
```
public function __construct() {
}
```

sharidas replied:
Done

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,262 @@
+			));
+		$statement = $qb->execute();
+		while ($row = $statement->fetch()) {
+			$whitelistedApps = \array_unique(\array_merge($whitelistedApps, \explode(',', $row['apps'])));
```
```
foreach (\explode(',', $row['apps']) as $app) {
  $whitelistedApps[$app] = $app;
}
```
That's likely faster. You don't need to make the array unique per row.
You can also return that kind of array (as long as it's documented) to avoid calling `array_values` to generate a new array.

sharidas replied:
Done. Updated it.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,262 @@
+				)
+			));
+		$statement = $qb->execute();
+		while ($row = $statement->fetch()) {
```
double-check this. Most of the code examples uses`while (($row = $statement->fetch()) !== false)`.

sharidas replied:
Most of the code in core is following `if/while` `($row = $statment->fetch())`. Hence I followed this.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,262 @@
+		}
+		$statement->closeCursor();
+
+		if (!empty($whitelistedApps)) {
```
Empty whitelist should be a possible result.
There are a couple of scenarios to check:
* If the user (or group) doesn't have a DB entry, we should return false.
* If the user (or group) has an entry with an empty list, we should return an empty list

sharidas replied:
So now if the list is empty returend by user or group we return empty list.
if user or group does not have DB entry, user has access to all the apps.

jvillafanez replied:
empty list should be a valid result

jvillafanez commented on core/Migrations/Version20191030124642.php:
```diff
@@ -0,0 +1,90 @@
+
+ 			$appAccessTable->addColumn(
+				'apps',
+				Type::BLOB,
```
Question: BLOB or TEXT? It seems to me that BLOB is inappropriate for this....
Alternatively, we could store the list into multiple rows, although this will require more code changes:
```
user1 | 0 | app1
user1 | 0 | app2
user1 | 0 | app3
....
```

sharidas replied:
Updated with TEXT.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,262 @@
+			->where($qb->expr()->eq('accessor', $qb->expr()->literal($user->getUID())))
+			->andWhere($qb->expr()->eq('accessoridentifier', $qb->expr()->literal(0)));
+		$statement = $qb->execute();
+		$result = $statement->fetch();
```
Include a comment saying that we expect only one row to be present for that user.

sharidas replied:
Added comment.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,262 @@
+			->where($qb->expr()->eq('accessor', $qb->expr()->literal($group->getGID())))
+			->andWhere($qb->expr()->eq('accessoridentifier', $qb->expr()->literal('1')));
+		$statement = $qb->execute();
+		$result = $statement->fetch();
```
Same as above

sharidas replied:
Done

jvillafanez commented on lib/public/App/IAppAccessService.php:
```diff
@@ -0,0 +1,118 @@
+	 * @param string[] $apps
+	 * @since 10.3.1
+	 */
+	public function setWhitelistedAppsForUsers(IUser $users, $apps = []);
```
We can remove this method. I don't think it adds any value. The caller can loop through the users and call the `setWhitelistedAppsForUser` method

sharidas replied:
Removed this method.

sharidas commented on core/Migrations/Version20191030124642.php:
```diff
@@ -0,0 +1,90 @@
+
+ 			$appAccessTable->setPrimaryKey(['id']);
+
+ 			$appAccessTable->addIndex(['accessor'], 'usergroupwith_whitelistedapps');
```
Done. Updated with `$appAccessTable->addIndex(['accessor', 'accessoridentifier'], 'usergroupwith_whitelistedapps');`

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,277 @@
+		$groupObjects = $this->groupManager->getUserGroups($user);
+		$whitelistedApps = [];
+		$grantAllAppsAccessForUser = true;
+		$userIds[] = $user->getUID();
```
This isn't used

sharidas replied:
Removed.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,277 @@
+		}
+
+		/**
+		 * First check if the user is available in the table.
```
This isn't needed. There are 3 outcomes with the next query (the one getting the list of apps):
* No results -> this means all apps are allowed
* Results combined to an empty list -> this means the user won't have any app whitelisted
* Results combined to a list -> the user will have only those apps whitelisted.

sharidas replied:
Removed and updated the code.

jvillafanez commented on lib/public/App/IAppAccessService.php:
```diff
@@ -0,0 +1,106 @@
+	 * @return bool true if whitelisted apps are set else false is returned
+	 * @since 10.3.1
+	 */
+	public function setWhitelistedAppsForGroup(IGroup $group, $apps = []);
```
Keep the method interface consistent. The `setWhitelistedAppsForUser` call requires an array of apps, but it's optional here.

sharidas replied:
I will update `setWhitelistedAppsForUser` to have optional. That way it becomes consistent.

jvillafanez commented:
It seems we missed something to remove the DB entries.

sharidas commented:
> It seems we missed something to remove the DB entries.

I have added 2 new methods:
- `grantAllAppsAccessForUser` -> to delete the user entry in the table and
- `grantAllAppsAccessForGroup` -> to delete the group entry in the table.

jvillafanez commented on lib/private/App/AppManager.php:
```diff
@@ -768,4 +785,18 @@ public function getAppVersionByPath($path) {
+	 * @since 10.3.1
+	 */
+	public function getAppsAccessServiceForUser() {
+		$this->dispatcher->addListener('group.postDelete', function (GenericEvent $event) {
```
this is a really bad place to add hook. Each call will add a new hook.
In addition, try to keep all the hooks as close as possible (although it isn't a hook, you're doing things when the user is deleted), otherwise it will be difficult to find things.

sharidas replied:
Moved this change to Server. Both the hook and event are now in the same place. Also its a small change of code. I have also removed the change made to User.php, where I called the wipe method for user in the delete method.

sharidas commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,310 @@
+
+	/**
+	 * Delete association of whitelisted apps and user
+	 * This method will grant full app access to the user as after this operation
```
Done.

sharidas commented:
Updated the changes requested for rename of methods mentioned here https://github.com/owncloud/core/pull/36338#issuecomment-551442545

sharidas commented:
Tested the sections:
- Created a user `adminSubst` and added this user to `admin` group
- Add empty list as whitelisted apps for this user ( just create a small script to call the API in this pr).
- Now try to navigate to the settings page
- The sharing page and encryption page are visible in the admin panel
   - The encryption page, though might look ugly. But as of now it looks ok. Because the ui is fetched from the core( the template ) and not from the encryption app.
    - The sharing page also fetches the template for the ui from the core.

Also verified with encryption ( **masterkey** ):
- The `admin` user ( who has access to the encryption module ) can upload/download file. The file is encrypted in oC.
- The `adminSubst` can upload/download file. The file is not encrypted ( expected behaviour) , as encryption app is not whitelisted for the user.

jvillafanez commented on tests/lib/App/AppAccessServiceTest.php:
```diff
@@ -0,0 +1,165 @@
+	 * This test proves that if user is not there in the table, then the user
+	 * has access to all the apps.
+	 */
+	public function testGetWhitelistedAppsForUserNotAddedToTable() {
```
Having a `getWhitelistedAppsForUser` method I'd expect this test to test that method, not the "computed" one. Adjust the name of the test to reflect what you're testing.

jvillafanez commented on tests/lib/App/AppAccessServiceTest.php:
```diff
@@ -0,0 +1,165 @@
+		$this->appAccessService->setWhitelistedAppsForUser($this->user2, ['appUser2', 'appUser20']);
+
+		$this->assertEquals(
+			['app1' => 'app1', 'app2' => 'app2', 'app3' => 'app3', 'app4' => 'app4',
```
```
$expectedAppList = [
  'app1' => 'app1',
  'app2' => 'app2',
  ....
];
$this->assertEquals($expectedAppList, $this->appAccessService->.....);
```
This seems more readable

jvillafanez commented on tests/lib/App/DummyAppAccessServiceTest.php:
```diff
@@ -0,0 +1,72 @@
+		$this->assertFalse($this->dummyAppAccessService->setWhitelistedAppsForUser($user, []));
+	}
+
+	public function testGetWhitelistedAppsForUser() {
```
Adjust the name of the test

jvillafanez commented on tests/lib/App/ManagerTest.php:
```diff
@@ -405,10 +414,27 @@ public function testGetAppsForUser() {
 		], $this->manager->getEnabledAppsForUser($user));
 	}
 
+	public function testGetWhitelistedAppsForUser() {
```
change the name to `testGetEnabledAppsForUserWithWhitelist` or similar to make more clear what you're testing.

jvillafanez commented on lib/private/App/AppManager.php:
```diff
@@ -160,9 +166,13 @@ public function getInstalledApps() {
-		});
+		$regulatedAppForUser = false;
+		if ($user !== null) {
+			$regulatedAppForUser = $this->getAppsAccessServiceForUser()->getWhitelistedAppsForUser($user);
```
:exclamation: This should be using the "computed" version, otherwise groups won't be considered.

jvillafanez commented on lib/private/App/AppAccessService.php:
```diff
@@ -0,0 +1,329 @@
+	 * - If the user is part of a group, and if that group has whitelisted apps,
+	 *   then the user still obeys the whitelisted apps of a group ( which it is a
+	 *   member of)
+	 *   member of)
```
duplicated line

sharidas replied:
Fixed.

jvillafanez commented on tests/lib/App/AppAccessServiceTest.php:
```diff
@@ -0,0 +1,300 @@
+		$this->appAccessService->setWhitelistedAppsForGroup($this->group2, ['hello', 'world']);
+
+		$result = $this->appAccessService->getWhitelistedAppsForGroup($this->group1);
+		$this->assertEquals(['foo', 'bar'], $result, 'Ignore the order of the apps', 0.0, 10, true);
```
Don't use undocumented parameters.

sharidas replied:
Removed this usage. Found another way to tackle the problem.

jvillafanez reviewed (commented):
Last details for me

jvillafanez commented on tests/lib/App/AppAccessServiceTest.php:
```diff
@@ -0,0 +1,285 @@
+			'app5' => 'app5',
+			'appUser1' => 'appUser1',
+			'appUser10' => 'appUser10'
+		], $this->appAccessService->getComputedWhitelistedAppsForUser($this->user1)));
```
This seems badly formatted, and could be confusing when reading it. There are other cases above and below.

sharidas commented:
@DeepDiver1975 I would like your review on this PR. @jvillafanez had reviewed this PR. I would like your opinion, as its a new API change to core. Thanks.
